crates/core/tedge_mapper/src/c8y/converter.rs

```diff
@@ -224,31 +226,79 @@ where
-            let json_mqtt_topic = Topic::new_unchecked(C8Y_JSON_MQTT_EVENTS_TOPIC);
+                c8y_event = C8yCreateEvent::try_from(tedge_event)?;
+                // If the message contains extra fields other than `text` and `time`, convert to Cumulocity JSON
+                let cumulocity_event_json = serde_json::to_string(&c8y_event)?;
 
-            Message::new(&json_mqtt_topic, cumulocity_event_json)
+                Message::new(
+                    &Topic::new_unchecked(C8Y_JSON_MQTT_EVENTS_TOPIC),
+                    cumulocity_event_json,
+                )
+            }
+            None => {
+                let tedge_event =
+                    ThinEdgeEvent::try_from(input.topic.name.as_str(), input.payload_str()?, None)?;
+                c8y_event = C8yCreateEvent::try_from(tedge_event)?;
+
+                // If the message doesn't contain any fields other than `text` and `time`, convert to SmartREST
+                if c8y_event.extras.is_empty() {
+                    let smartrest_event = Self::serialize_to_smartrest(&c8y_event)?;
+                    let smartrest_topic = Topic::new_unchecked(SMARTREST_PUBLISH_TOPIC);
+
+                    Message::new(&smartrest_topic, smartrest_event)
+                } else {
+                    // If the message contains extra fields other than `text` and `time`, convert to Cumulocity JSON
+                    let cumulocity_event_json = serde_json::to_string(&c8y_event)?;
+                    let json_mqtt_topic = Topic::new_unchecked(C8Y_JSON_MQTT_EVENTS_TOPIC);
+
+                    Message::new(&json_mqtt_topic, cumulocity_event_json)
+                }
+            }
         };
 
         // If the MQTT message size is well within the Cumulocity MQTT size limit, use MQTT to send the mapped event as well
         if input.payload_bytes().len() < self.size_threshold.0 {
-            Ok(vec![message])
+            vec.push(message)
+        // If the message size is larger than the MQTT size limit, use HTTP to send the mapped event
         } else {
-            // If the message size is larger than the MQTT size limit, use HTTP to send the mapped event
-            let _ = self.http_proxy.send_event(c8y_event).await?;
-            Ok(vec![])
+            match child_id {
+                Some(id) => {
+                    if self.children.contains(&id) {
+                        let _ = self.http_proxy.send_event(c8y_event).await?;
+                        return Ok(vec![]);
+                    } else {
+                        return Err(ConversionError::ChildDeviceNotRegistered {
+                            id: id.to_string(),
+                        });
+                    }
+                }
+                // Parent device
+                None => {
+                    let _ = self.http_proxy.send_event(c8y_event).await?;
+                    return Ok(vec![]);
+                }
+            }
         }
+
+        Ok(vec)
```

Comment:
Move this `Ok(vec)` to after line 279. Then you can remove all `return` from the `else` block.

```rust
        if input.payload_bytes().len() < self.size_threshold.0 {
            vec.push(message)
            Ok(vec)
       } else {
       // remove all `return` from this block
       }
```

Reply:
Fixed